### PR TITLE
petsc: 3.8.4 -> 3.10.3

### PR DESCRIPTION
**Pull request by jamtrott**
These changes upgrade PETSc to the most recent release and add several useful linear solver backends, including HYPRE, MUMPS, scalapack, and SPAI.  In addition, the Python wrapper package petsc4py is built.

This has been tested on Mac OS X 10.13.6 (High Sierra).

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/libraries/science/math/hypre/default.nix
Comment (srhb) [inline pkgs/development/libraries/science/math/hypre/default.nix]: `stdenv.lib`

File: pkgs/development/libraries/science/math/mumps/default.nix
Comment (srhb) [inline pkgs/development/libraries/science/math/mumps/default.nix]: `stdenv.lib`

File: pkgs/development/libraries/science/math/petsc/default.nix
Comment (srhb) [inline pkgs/development/libraries/science/math/petsc/default.nix]: `stdenv.lib`

File: pkgs/development/libraries/science/math/spai/default.nix
Comment (srhb) [inline pkgs/development/libraries/science/math/spai/default.nix]: `stdenv.lib` for both license and maintainers.

File: pkgs/development/python-modules/petsc4py/default.nix
Comment (srhb) [inline pkgs/development/python-modules/petsc4py/default.nix]: `stdenv.lib`

**Review by srhb — CHANGES_REQUESTED** [review]
Evaluation currently fails due to some misses in the meta sections. Can you fix these up? :)

**Comment by jamtrott** [comment]
> Evaluation currently fails due to some misses in the meta sections. Can you fix these up? :)

Hi Sarah. Thank you for reviewing!

I fixed those mistakes and have pushed some new commits. I'm not sure whether it was necessary to create a commit for each separate package, but they can always be squashed before merging anyway.

**Comment by srhb** [comment]
Ideally I'd like this to be squashed down to one commit per package, would you mind doing that? :)

**Comment by jamtrott** [comment]
> Ideally I'd like this to be squashed down to one commit per package, would you mind doing that? :)

No problem at all. What would be the correct way of doing so? Do I close this pull request and submit a new one? Or is it possible to overwrite the branch with a forced push?

**Comment by srhb** [comment]
@jamtrott The easiest way is the locally undo your changes with a soft reset, eg `git reset master` -- this will leave your changed files but uncommitted. You can now piecemeal commit each logical package, and finally force push your branch, which will update this PR. :)

File: pkgs/development/python-modules/mpi4py/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/mpi4py/default.nix]: Can you explain what `build_src --force does`? Maybe a comment inside the expression would be nice.

File: pkgs/development/python-modules/petsc4py/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/petsc4py/default.nix]:
```suggestion
    license = lib.licenses.bsd3;
```

File: pkgs/development/python-modules/petsc4py/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/petsc4py/default.nix]:
```suggestion
{ lib
```

File: pkgs/development/python-modules/petsc4py/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/petsc4py/default.nix]: Some of those should probably be propagated. Why is python there?

**Review by dotlambda — COMMENTED** [review]
Please add the `pythonPackages.` prefix to the respective commits.

File: pkgs/development/python-modules/petsc4py/default.nix
Comment (jamtrott) [reply pkgs/development/python-modules/petsc4py/default.nix]: Hi @dotlambda, thank you for your comments. I admit that I'm not sure about buildInputs versus propagatedBuildInputs. Do you have any suggestions or guidelines for determining which packages should be propagated?

The python package is not needed, and I have removed it from the list.

Comment (dotlambda) [reply pkgs/development/python-modules/petsc4py/default.nix]: See https://nixos.org/nixpkgs/manual/#reference:

> If something is exclusively a build-time dependency, use buildInputs; if it is (also) a runtime dependency, use propagatedBuildInputs.

**Comment by jamtrott** [comment]
@srhb, @dotlambda, would you agree that the latest revisions address the issues you have raised?

File: pkgs/development/python-modules/petsc4py/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/petsc4py/default.nix]: You should add the same comment here.

File: pkgs/development/python-modules/petsc4py/default.nix
Comment (dotlambda) [reply pkgs/development/python-modules/petsc4py/default.nix]: See above for `cython`.
I don't find `mpi4py` in their [setup.py](https://bitbucket.org/petsc/petsc4py/src/6fe2748c55f524f51f93dc8d119d650fee44f9d0/setup.py).
`numpy` is in `install_requires` and should thus be propagated.
What do you infer the dependency on openssh from?

File: pkgs/development/python-modules/petsc4py/default.nix
Comment (dotlambda) [inline pkgs/development/python-modules/petsc4py/default.nix]: I don't find `openmpi` in ther setup.py.
`petsc` can probably be dropped from here since you specify `PETSC_DIR`.

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/mpi4py/default.nix
Comment (jamtrott) [reply pkgs/development/python-modules/mpi4py/default.nix]: Thanks, this should be fixed now.

File: pkgs/development/python-modules/petsc4py/default.nix
Comment (jamtrott) [reply pkgs/development/python-modules/petsc4py/default.nix]: Done.

File: pkgs/development/python-modules/petsc4py/default.nix
Comment (jamtrott) [reply pkgs/development/python-modules/petsc4py/default.nix]: I moved `openmpi` to `nativeBuildInputs` and removed `petsc`.  The package doesn't build without `openmpi`.

**Comment by tkonolige** [comment]
Would it be possible to parametrize these packages by which MPI implementation they use instead of hardcoding openmpi?

**Comment by FRidh** [comment]
Would be good to see the merge conflicts resolved, because this seems otherwise ready.

**Pull request closed without merging**
